### PR TITLE
Make `ptr::eq` documentation mention fat-pointer behavior

### LINKED ISSUE #59214: [docs] `ptr::eq` source does not mention smart-pointer comparison behavior.
The documentation at the [`ptr::eq` source] doesn't mention that smart pointers are compared in their totality, unlike pointer comparison which just compares the addresses. I found out about this behavior on [this urlo question](https://users.rust-lang.org/t/shallow-compare-for-slices/26321/3).

[`ptr::eq` source]: https://github.com/rust-lang/rust/blob/master/src/libcore/ptr.rs#L2476

### PR DESCRIPTION
**Pull request by czipperz**
Resolves #59214

### CODE REVIEW

**Comment by CryZe** [comment]
I'm pretty sure this is incorrect. The difference is that *fat* pointers (not smart pointers) compare their metadata in addition to the pointer address.

**Comment by steveklabnik** [comment]
Yes, I agree with @cryze

**Comment by czipperz** [comment]
What do you mean by a fat pointer?  Why is smart pointer wrong?

The three example smart pointers I listed compare by value. They use compare by address only as an optimization when the type implements Eq according to reading the source code.

**Comment by czipperz** [comment]
Googled fat pointers:

* slices with length
* traits with vtables

**Comment by czipperz** [comment]
KK so the issue is wrong. I'll fix this. Thanks guys

**Comment by czipperz** [comment]
Should we be adding a method that compares pointers by purely memory address without comparing metadata?

**Comment by czipperz** [comment]
@steveklabnik

**Comment by steveklabnik** [comment]
@bors: r+ rollup

thank you!

File: src/libcore/ptr.rs [hunk 1]
```diff
@@ -2495,11 +2495,57 @@ impl<T: ?Sized> Eq for *mut T {}
...
+/// struct Wrapper { member: i32 }
+///
+/// trait Trait {}
+/// impl Trait for Wrapper {}
+/// impl Trait for i32 {}
+///
+/// fn main() {
+///     let wrapper = Wrapper { member: 10 };
+///
+///     // Pointers are equal address
```
Comment (Centril) [inline src/libcore/ptr.rs]:
```suggestion
///     // Pointers have equal addresses.
```

File: src/libcore/ptr.rs [hunk 2]
```diff
@@ -2495,11 +2495,57 @@ impl<T: ?Sized> Eq for *mut T {}
...
+///     assert!(std::ptr::eq(
+///         &wrapper as *const Wrapper as *const u8,
+///         &wrapper.member as *const i32 as *const u8
+///     ));
+///
+///     // Objects have equal addresses, but `Trait` has different implementations
```
Comment (Centril) [inline src/libcore/ptr.rs]:
```suggestion
///     // Objects have equal addresses, but `Trait` has different implementations.
```

File: src/libcore/ptr.rs [hunk 3]
```diff
@@ -2495,11 +2495,57 @@ impl<T: ?Sized> Eq for *mut T {}
...
+///     assert!(!std::ptr::eq(
+///         &wrapper as &Trait,
+///         &wrapper.member as &Trait,
+///     ));
+///     assert!(!std::ptr::eq(
+///         &wrapper as &Trait as *const Trait,
+///         &wrapper.member as &Trait as *const Trait,
+///     ));
+///
+///     // Converting the reference to a `*const u8` compares by address
```
Comment (Centril) [inline src/libcore/ptr.rs]:
```suggestion
///     // Converting the reference to a `*const u8` compares by address.
```

File: src/libcore/ptr.rs [hunk 4]
```diff
@@ -2495,11 +2495,57 @@ impl<T: ?Sized> Eq for *mut T {}
...
+///
+///     // Pointers are equal address
+///     assert!(std::ptr::eq(
+///         &wrapper as *const Wrapper as *const u8,
+///         &wrapper.member as *const i32 as *const u8
+///     ));
+///
+///     // Objects have equal addresses, but `Trait` has different implementations
+///     assert!(!std::ptr::eq(
+///         &wrapper as &Trait,
```
Comment (Centril) [inline src/libcore/ptr.rs]:
```suggestion
///         &wrapper as &dyn Trait,
```

File: src/libcore/ptr.rs [hunk 5]
```diff
@@ -2495,11 +2495,57 @@ impl<T: ?Sized> Eq for *mut T {}
...
+///         &wrapper.member as &Trait,
```
Comment (Centril) [inline src/libcore/ptr.rs]:
```suggestion
///         &wrapper.member as &dyn Trait,
```

File: src/libcore/ptr.rs [hunk 6]
```diff
@@ -2495,11 +2495,57 @@ impl<T: ?Sized> Eq for *mut T {}
...
+///     ));
+///     assert!(!std::ptr::eq(
+///         &wrapper as &Trait as *const Trait,
```
Comment (Centril) [inline src/libcore/ptr.rs]:
```suggestion
///         &wrapper as &dyn Trait as *const dyn Trait,
```

File: src/libcore/ptr.rs [hunk 7]
```diff
@@ -2495,11 +2495,57 @@ impl<T: ?Sized> Eq for *mut T {}
...
+///         &wrapper.member as &Trait as *const Trait,
```
Comment (Centril) [inline src/libcore/ptr.rs]:
```suggestion
///         &wrapper.member as &dyn Trait as *const dyn Trait,
```

File: src/libcore/ptr.rs [hunk 8]
```diff
@@ -2495,11 +2495,57 @@ impl<T: ?Sized> Eq for *mut T {}
...
+///     ));
+///
+///     // Converting the reference to a `*const u8` compares by address
+///     assert!(std::ptr::eq(
+///         &wrapper as &Trait as *const Trait as *const u8,
```
Comment (Centril) [inline src/libcore/ptr.rs]:
```suggestion
///         &wrapper as &dyn Trait as *const dyn Trait as *const u8,
```

File: src/libcore/ptr.rs [hunk 9]
```diff
@@ -2495,11 +2495,57 @@ impl<T: ?Sized> Eq for *mut T {}
...
+///         &wrapper.member as &Trait as *const Trait as *const u8,
```
Comment (Centril) [inline src/libcore/ptr.rs]:
```suggestion
///         &wrapper.member as &dyn Trait as *const dyn Trait as *const u8,
```

**Review by Centril — CHANGES_REQUESTED** [review]
We shouldn't be using `Trait` instead of `dyn Trait` anymore in the standard library.

**Comment by czipperz** [comment]
KK @Centril  done

**Review by Centril — APPROVED** [review]

**Comment by Centril** [comment]
r=me,steveklabnik with green travis

**Comment by czipperz** [comment]
@Centril @steveklabnik

**Pull request merged**